tools/run-fbp-tests

```diff
@@ -44,6 +44,104 @@ import re
 import subprocess
 import sys
 
+valgrind_opts = [
+    "--show-reachable=no",
+    "--quiet",
+    "--error-exitcode=1",
+    "--leak-check=full",
+]
+
+class FbpTest:
+    SKIP_NONE = 0
+    SKIP_SIMPLE = 1
+    SKIP_VALGRIND = 2
+    SKIP_COMPILE = 3
+
+    supported_skip = {
+        "TEST-SKIP": SKIP_SIMPLE,
+        "TEST-SKIP-VALGRIND": SKIP_VALGRIND,
+        "TEST-SKIP-COMPILE": SKIP_COMPILE,
+    }
+
+    def __init__(self, test_file):
+        self.test_file = test_file
+        self.skip_msg = None
+        self.skip = self.SKIP_NONE
+        self.output = False
+        self.output_regex = False
+        self.output_spec = ""
+        self.exit_code = 0
+
+        self.__parse_test__()
+
+    def __parse_skip__(self, cmd, splitted):
+        self.skip = self.supported_skip.get(cmd, self.SKIP_NONE)
+        if not self.skip:
+            return
+
+        if len(splitted) > 2:
+            self.skip_msg = splitted[2]
+
+    def __parse_command__(self, line):
+        splitted = line.split(None, 2)
+        cmd = splitted[1]
+
+        if cmd == "TEST-OUTPUT-REGEX":
+            self.output_regex = True
+        elif cmd == "TEST-OUTPUT":
+            self.output = True
+        elif cmd == "TEST-EXPECTS-ERROR":
+            self.exit_code = 1
+        elif cmd.startswith("TEST-SKIP"):
+            self.__parse_skip__(cmd, splitted)
+        else:
+            print("Ignoring unknown test command '%s' in test %s" %
+                  (cmd, self.test_file))
+
+    def __parse_test__(self):
+        f = open(self.test_file, encoding="UTF-8")
+        for line in f:
```

Comment:
Nitpick, but this can be written like so:

     for line in open(...):
         do_stuff_with(line)